sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -543,14 +544,52 @@ RT::PiProgram ProgramManager::getBuiltPIProgram(
 
   DeviceImplPtr Dev =
       (MustBuildOnSubdevice == PI_TRUE) ? DeviceImpl : RootDevImpl;
-  auto BuildF = [this, &M, &KSId, &ContextImpl, &Dev, Prg, &CompileOpts,
-                 &LinkOpts, &JITCompilationIsRequired, SpecConsts] {
-    auto Context = createSyclObjFromImpl<context>(ContextImpl);
-    auto Device = createSyclObjFromImpl<device>(Dev);
+  auto Context = createSyclObjFromImpl<context>(ContextImpl);
+  auto Device = createSyclObjFromImpl<device>(Dev);
+  const RTDeviceBinaryImage &Img =
+      getDeviceImage(M, KSId, Context, Device, JITCompilationIsRequired);
+
+  // Check that device supports all aspects used by the kernel
+  const RTDeviceBinaryImage::PropertyRange &ARange =
+      Img.getDeviceRequirements();
+
+#define __SYCL_ASPECT(A, B, C)                                                 \
+  case aspect::A:                                                              \
+    return C;
+#define __SYCL_SPECIAL_ASPECT(A, B, C)
+  auto getAspectNameStr = [](aspect AspectNum) -> std::string {
+    switch (AspectNum) {
+#include <sycl/info/aspects.def>
+    default:
+      throw sycl::exception(
+          errc::kernel_not_supported,
+          "Unknown aspect " + std::to_string(static_cast<unsigned>(AspectNum)));
+    }
+  };
+#undef __SYCL_SPECIAL_ASPECT
+#undef __SYCL_ASPECT
 
-    const RTDeviceBinaryImage &Img =
-        getDeviceImage(M, KSId, Context, Device, JITCompilationIsRequired);
+  for (RTDeviceBinaryImage::PropertyRange::ConstIterator It : ARange) {
+    auto KName = std::string((*It)->Name);
```

Comment:
We have switched to C++17. Would `std::string_view` suffice here?

Reply:
Okay, https://github.com/intel/llvm/pull/6989/commits/410f00eabc04a1ab5be42dc4da11c20fd07d51ae